File: agents/backend/onyx/server/features/addiction_recovery_ai/addiction_recovery_ai_mobile/src/utils/security.ts
import { sanitizeString } from './sanitize';

export function validateAndSanitizeInput(
  input: string,
  maxLength?: number
): { isValid: boolean; sanitized: string; error?: string } {
  if (!input || typeof input !== 'string') {
    return {
      isValid: false,
      sanitized: '',
      error: 'Input must be a non-empty string',
    };
  }

  if (maxLength && input.length > maxLength) {
    return {
      isValid: false,
      sanitized: '',
      error: `Input must be less than ${maxLength} characters`,
    };
  }

  const sanitized = sanitizeString(input);

  if (sanitized.length === 0) {
    return {
      isValid: false,
      sanitized: '',
      error: 'Input contains only invalid characters',
    };
  }

  return {
    isValid: true,
    sanitized,
  };
}

export function validateEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

export function validateURL(url: string): boolean {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

export function maskSensitiveData(data: string, visibleChars = 4): string {
  if (data.length <= visibleChars) {
    return '*'.repeat(data.length);
  }

  const visible = data.slice(-visibleChars);
  const masked = '*'.repeat(data.length - visibleChars);
  return masked + visible;
}

export function generateSecureToken(length = 32): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let token = '';
  
  for (let i = 0; i < length; i++) {
    token += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  
  return token;
}

File: agents/backend/onyx/server/features/addiction_recovery_ai/utils/security.py
"""
Security utilities
"""

import hashlib
import secrets
from typing import Optional
import re


def generate_secure_token(length: int = 32) -> str:
    """
    Generate a cryptographically secure random token
    
    Args:
        length: Length of token in bytes
    
    Returns:
        Hexadecimal token string
    """
    return secrets.token_hex(length)


def hash_string(value: str, algorithm: str = "sha256") -> str:
    """
    Hash a string using specified algorithm
    
    Args:
        value: String to hash
        algorithm: Hash algorithm (md5, sha1, sha256, sha512)
    
    Returns:
        Hexadecimal hash string
    """
    hash_algorithms = {
        "md5": hashlib.md5,
        "sha1": hashlib.sha1,
        "sha256": hashlib.sha256,
        "sha512": hashlib.sha512
    }
    
    hash_func = hash_algorithms.get(algorithm.lower(), hashlib.sha256)
    return hash_func(value.encode()).hexdigest()


def validate_password_strength(password: str) -> tuple[bool, list[str]]:
    """
    Validate password strength
    
    Args:
        password: Password to validate
    
    Returns:
        Tuple of (is_valid, list_of_issues)
    """
    issues = []
    
    if len(password) < 8:
        issues.append("Password must be at least 8 characters long")
    
    if not re.search(r'[A-Z]', password):
        issues.append("Password must contain at least one uppercase letter")
    
    if not re.search(r'[a-z]', password):
        issues.append("Password must contain at least one lowercase letter")
    
    if not re.search(r'\d', password):
        issues.append("Password must contain at least one digit")
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        issues.append("Password must contain at least one special character")
    
    return len(issues) == 0, issues


def sanitize_input(value: str, max_length: Optional[int] = None) -> str:
    """
    Sanitize user input
    
    Args:
        value: Input string to sanitize
        max_length: Maximum allowed length
    
    Returns:
        Sanitized string
    """
    if not isinstance(value, str):
        raise ValueError("Input must be a string")
    
    # Remove leading/trailing whitespace
    sanitized = value.strip()
    
    # Remove null bytes
    sanitized = sanitized.replace('\x00', '')
    
    # Limit length
    if max_length and len(sanitized) > max_length:
        sanitized = sanitized[:max_length]
    
    return sanitized


def mask_sensitive_data(value: str, visible_chars: int = 4) -> str:
    """
    Mask sensitive data (e.g., email, phone)
    
    Args:
        value: Value to mask
        visible_chars: Number of characters to keep visible
    
    Returns:
        Masked string
    """
    if not value or len(value) <= visible_chars:
        return "*" * len(value) if value else ""
    
    return value[:visible_chars] + "*" * (len(value) - visible_chars)
